py-polars/polars/dataframe/frame.py

```diff
@@ -3405,15 +3409,27 @@ def check_unsupported_types(n: str, t: PolarsDataType | None) -> None:
 
         data = self.to_arrow()
         data_schema = data.schema
+        delta_schema = _create_delta_compatible_schema(data_schema)
 
         # Workaround to prevent manual casting of large types
         table = try_get_deltatable(target, storage_options)  # type: ignore[arg-type]
 
-        if table is not None:
-            table_schema = table.schema()
+        if data_schema != delta_schema:
+            data_schema = delta_schema
+
+            if table is not None:
+                table_schema = table.schema()
+
+                if data_schema == table_schema.to_pyarrow(as_large_types=True):
+                    data_schema = table_schema.to_pyarrow()
+            else:
+                data = data.cast(data_schema)
+        else:
+            if table is not None:
+                table_schema = table.schema()
 
-            if data_schema == table_schema.to_pyarrow(as_large_types=True):
-                data_schema = table_schema.to_pyarrow()
+                if data_schema == table_schema.to_pyarrow(as_large_types=True):
+                    data_schema = table_schema.to_pyarrow()
```

Comment:
There must be a better way to write this 🤔

Reply:
> There must be a better way to write this 🤔

I'll have to think about it! Will try to work on it tonight

Reply:
I've removed this part completely now since it became obsolete due to also reconstructing schema from large types to normal types.